src/main/java/org/opensearch/geospatial/index/mapper/xypoint/XYPointParser.java

```diff
@@ -71,101 +78,156 @@ public static XYPoint parseXYPoint(Object value, final boolean ignoreZValue) thr
+                parseXYPointArray(parser, point, ignoreZValue);
+                break;
+            case VALUE_STRING:
+                String val = parser.text();
+                point.resetFromString(val, ignoreZValue);
+                break;
+            default:
+                throw new OpenSearchParseException("xy_point expected");
+        }
+        return point;
+    }
+
+    private static XYPoint parseXYPointObject(final XContentParser parser, final XYPoint point, final boolean ignoreZValue)
+        throws IOException {
+        try (XContentSubParser subParser = new XContentSubParser(parser)) {
+            if (subParser.nextToken() != XContentParser.Token.FIELD_NAME) {
+                throw new OpenSearchParseException(ERR_MSG_INVALID_TOKEN, subParser.currentToken());
             }
-            if (Double.isNaN(x)) {
-                throw new OpenSearchParseException("field [{}] missing", X_PARAMETER);
+
+            String field = subParser.currentName();
+            if (X_PARAMETER.equals(field) || Y_PARAMETER.equals(field)) {
+                parseGeoPointObjectBasicFields(subParser, point);
+            } else if (GEOJSON_TYPE.equals(field) || GEOJSON_COORDS.equals(field)) {
+                parseGeoJsonFields(subParser, point, ignoreZValue);
+            } else {
+                throw new OpenSearchParseException(ERR_MSG_INVALID_FIELDS);
             }
-            if (Double.isNaN(y)) {
-                throw new OpenSearchParseException("field [{}] missing", Y_PARAMETER);
+
+            if (subParser.nextToken() != XContentParser.Token.END_OBJECT) {
+                throw new OpenSearchParseException(ERR_MSG_INVALID_FIELDS);
             }
-            return point.reset(x, y);
+
+            return point;
         }
+    }
+
+    private static XYPoint parseGeoPointObjectBasicFields(final XContentParser parser, final XYPoint point) throws IOException {
+        HashMap<String, Double> data = new HashMap<>();
+        for (int i = 0; i < 2; i++) {
+            if (i != 0) {
+                parser.nextToken();
+            }
+
+            if (parser.currentToken() != XContentParser.Token.FIELD_NAME) {
+                break;
+            }
 
-        if (parser.currentToken() == XContentParser.Token.START_ARRAY) {
-            return parseXYPointArray(parser, ignoreZValue, x, y);
+            String field = parser.currentName();
+            if (X_PARAMETER.equals(field) == false && Y_PARAMETER.equals(field) == false) {
+                throw new OpenSearchParseException(ERR_MSG_INVALID_FIELDS);
+            }
+            switch (parser.nextToken()) {
+                case VALUE_NUMBER:
+                case VALUE_STRING:
+                    try {
+                        data.put(field, parser.doubleValue(true));
+                    } catch (NumberFormatException e) {
+                        throw new OpenSearchParseException("[{}] and [{}] must be valid double values", e, X_PARAMETER, Y_PARAMETER);
```

Comment:
The message is used only in this method. I don't think it is worth to make is as a constant?

Reply:
I'm leaving it up to you, but I would make it a constant anyway. From my experience it helps to avoid creating multiple identical messages with slightly different wording. If all messages are in one place chance is higher that next person will reuse one from the list.